### PR TITLE
Don't delete gru task from db if minion job has not completed

### PR DESCRIPTION
**Pull request by AdamWill**
In Fedora openQA, I noticed a problem related to the recent
change to run gru/minion tasks/jobs in parallel. We put guards
in download_asset and several other non-parallel-safe tasks to
prevent multiple instances of them running in parallel with
each other: the tasks use a 'guard' (basically a lock) and if
they cannot acquire it, they restart themselves.

However, it seems that when this happens, the minion execute()
returns and we wind up in this block of our superclassed
execute() which assumes that at this point the job has actually
run, and deletes the corresponding entry from the gru task db.
This was a big problem in Fedora, because we schedule the same
asset as multiple flavors, so each flavor gets its jobs blocked
on a download_asset gru task. One task executes 'normally' and
will not be deleted from the gru task db until the asset is
actually downloaded, but the other task hits the retry mechanism
and is immediately deleted from the db even though the asset has
not yet downloaded...so all the jobs for that flavor run, and
fail immediately.

To protect against this, we can check the state of the minion
job after we call `finish` and only delete the task from the db
if the job is actually 'failed' or 'finished'. Otherwise we
leave the entry in the db and wait for the job to run again and
actually complete this time.

Related ticket: [poo#48554](https://progress.opensuse.org/issues/48554)

Signed-off-by: Adam Williamson <awilliam@redhat.com>

### CODE REVIEW

**Comment by AdamWill** [comment]
So, this does seem to actually work to solve the bug I'm trying to solve. I have it running on Fedora openQA staging now and it seems to be behaving correctly.

There are a couple of questions, though:

1. Is it possible for the job state to be 'active' after `finish` is called? If so, should we delete the task from the db or not at that point?
2. Would it be better to use `$self->is_finished` instead of the string check on `$self->info->{state}`?

I really hope the answer to 1 is "it's not possible" because it's a really tricky choice to make. If we delete the task from the gru task db while the minion job is actually still running, that's clearly wrong...but if somehow we get here with the state as 'active', if we *don't* delete the job, can we be sure we'll wind up back here with the state as 'finished' or 'failed'? I am just not sure. I do *think* the answer is simply "it's not possible", but I'm not 100% sure.

As to 2 - `is_finished` basically [runs through `waitpid`](https://github.com/mojolicious/minion/blob/master/lib/Minion/Job.pm#L36-L41). I don't really know if that would be more reliable or less reliable here. Not sure if anyone has an opinion.

File: lib/OpenQA/WebAPI/GruJob.pm [hunk 1]
```diff
@@ -49,8 +50,23 @@ sub execute {
...
+        my $state = $self->info->{state};
+        my $jobid = $self->id;
+        my $gruid;
+        $gruid = $self->info->{notes}{gru_id} if exists $self->info->{notes}{gru_id};
+        log_debug("After execution and finish of minion job $jobid / gru task $gruid, state is $state");
+        # if the job is retried at this point, its state will be 'inactive'
+        # and we do *not* want to delete the corresponding entry from the gru
+        # task database. We only want to delete that if the job is really done
+        if (grep { /$state/ } ('failed', 'finished')) {
+            if ($gruid) {
```
Comment (AdamWill) [inline lib/OpenQA/WebAPI/GruJob.pm]: in case you're wondering why I nested this: `if (grep { /$state/ } ('failed', 'finished') && $gruid)` doesn't actually seem to *work* right for some reason. If anyone knows why and can fix it, let me know. :D

**Comment by AdamWill** [comment]
@Martchus

**Comment by AdamWill** [comment]
And yes, no tests yet. I am not sure how best to go about testing this, but I'll try and think of something if I can...

File: lib/OpenQA/WebAPI/GruJob.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/GruJob.pm]: My guess would be that due to precedence rules the last expression (`('failed', 'finished') && $gruid`) gets evaluated first.

**Comment by Martchus** [comment]
> There are a couple of questions, though

Maybe @kraih knows.

---

The change makes sense in general, but yes, tests are missing.

**Comment by kraih** [comment]
Unfortunately the answer to 1 is yes, Minion allows for a job to be retried again after it reached the state finished. This is not done in openQA to my knowledge, but it is possible for a user to retry a finished job from the Minion admin ui. Also, finished and failed to Minion are not the same. Minion never deletes a failed job automatically. The assumption is that a failed job has to be reviewed by a user or bot (who can then retry or remove it manually).

File: lib/OpenQA/WebAPI/GruJob.pm (hunk 1 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/GruJob.pm]: `if ((grep { /$state/ } ('failed', 'finished')) && $gruid)` should work though, right?

Comment (kraih) [reply lib/OpenQA/WebAPI/GruJob.pm]: Although `grep { $_ eq $state }` would be better style.

Comment (kraih) [reply lib/OpenQA/WebAPI/GruJob.pm]: Actually, why not be totally clear with `if (($state eq 'finished' || $state eq 'failed') && $gruid)`

File: lib/OpenQA/WebAPI/GruJob.pm [hunk 2]
```diff
@@ -17,6 +17,7 @@ package OpenQA::WebAPI::GruJob;
 use Mojo::Base 'Minion::Job';
 
 use OpenQA::Utils 'log_error';
+use OpenQA::Utils 'log_debug';
```
Comment (kraih) [inline lib/OpenQA/WebAPI/GruJob.pm]: `use OpenQA::Utils qw(log_debug log_error);`

File: lib/OpenQA/WebAPI/GruJob.pm [hunk 3]
```diff
@@ -49,8 +50,23 @@ sub execute {
...
     }
     else {
+        # here we call finish, which will either actually finish the job or
+        # retry it
         $self->finish(defined $buffer ? $buffer : 'Job successfully executed');
-        $self->_delete_gru($self->info->{notes}{gru_id}) if exists $self->info->{notes}{gru_id};
+        my $state = $self->info->{state};
+        my $jobid = $self->id;
+        my $gruid;
+        $gruid = $self->info->{notes}{gru_id} if exists $self->info->{notes}{gru_id};
```
Comment (kraih) [inline lib/OpenQA/WebAPI/GruJob.pm]: You're triple requesting `$self->info` here, that's two extra Postgres roundtrips for the whole info data structure.

**Comment by kraih** [comment]
And i agree, this problem definitely needs a test case to decide which would be the correct solution.

File: lib/OpenQA/WebAPI/GruJob.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/GruJob.pm]: I prefer the last version.

**Comment by kraih** [comment]
@AdamWill And regarding 2, `->is_finished` is only for Minion job process management and unrelated to the actual state of the job. So definitely a no on using it here.

**Comment by kraih** [comment]
Think i have an idea for how to test this, will open a separate pull request for it later today or tomorrow.

File: lib/OpenQA/WebAPI/GruJob.pm (hunk 1 above)
Comment (AdamWill) [reply lib/OpenQA/WebAPI/GruJob.pm]: ahhh, my old adversary the bracket, it is you once more?!

thanks guys, will tweak.

File: lib/OpenQA/WebAPI/GruJob.pm (hunk 3 above)
Comment (AdamWill) [reply lib/OpenQA/WebAPI/GruJob.pm]: oh, fair point. will fix.

**Comment by AdamWill** [comment]
> Unfortunately the answer to 1 is yes, Minion allows for a job to be retried again after it reached the state finished. This is not done in openQA to my knowledge, but it is possible for a user to retry a finished job from the Minion admin ui. Also, finished and failed to Minion are not the same. Minion never deletes a failed job automatically. The assumption is that a failed job has to be reviewed by a user or bot (who can then retry or remove it manually).
> 
> **Edit:** Since Gru hijacks the `execute` method of Minion there are actually many cases where the job will still be `active`. That means the proposed solution does not work correctly.

thanks, I'm not entirely sure I'm following the comment, though. Question 1 was about whether we can reach this specific point in the subclassed `execute()` with the job state as 'active' - just because a job is retried, why would we get here with its state as 'active'?

If you're also saying that we shouldn't really delete jobs from GruTasks even after they're 'finished' because they might be re-tried, then indeed that seems a valid concern, but then this PR isn't making things *worse* there - we're *already* doing that right now.

**Comment by AdamWill** [comment]
Concerning the "jobs can be retried under the same ID" case - well, what GruTasks really is these days (AIUI) is a table of *pending and active* jobs. From my poking around in the code, I get the impression that minion emits events, right? Could we perhaps redesign things so jobs get added to and removed from GruTasks based on the events emitted on state changes?

As it seems like it's going to be relevant: does anyone know of anything remaining in openQA that actually *uses* the GruTasks / GruDependencies tables besides this 'openQA tests can be wait to wait on gru tasks' mechanism?

**Comment by kraih** [comment]
@AdamWill Yes, i did not explain that very well, but the subject is rather complicated. 😉 Anyway, i do think i have a solution and am now thinking of more edge cases to test. https://github.com/os-autoinst/openQA/compare/master...kraih:gru_tasks_cleanup

**Comment by AdamWill** [comment]
OK, thanks :) I think it might be wise to preserve the checks around whether keys in the `{info}` hash exist at all? Presumably they're there for a reason (likely you get errors or ugly log messages if trying to use `$self-{info}->{gru_id}` if it doesn't exist at all).

**Comment by kraih** [comment]
@AdamWill Perl autovivifies that, no need to worry. Here you would only use `exists` if a valid Gru id could be a false Perl value, which i don't believe to be the case.

**Comment by kraih** [comment]
This should be resolved with #2011.
